generator/lib/src/dart_type_ex.dart

```diff
@@ -0,0 +1,112 @@
+////////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2021 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+import 'dart:ffi';
+import 'dart:typed_data';
+
+import 'package:analyzer/dart/element/type.dart';
+import 'package:realm_annotations/realm_annotations.dart';
+import 'package:source_gen/source_gen.dart';
+
+import 'session.dart';
+import 'type_checkers.dart';
+
+extension DartTypeEx on DartType {
+  bool isExactly<T>() => TypeChecker.fromRuntime(T).isExactlyType(this);
+
+  bool get isRealmAny =>
+      const TypeChecker.fromRuntime(RealmAny).isAssignableFromType(this);
+  bool get isRealmBacklink => false; // TODO
+  bool get isRealmCollection => realmCollectionType != RealmCollectionType.none;
+  bool get isRealmModel =>
+      realmModelChecker.annotationsOfExact(element!).isNotEmpty;
+
+  bool get isNullable => session.typeSystem.isNullable(this);
+  DartType get asNonNullable => session.typeSystem.promoteToNonNull(this);
+
+  RealmCollectionType get realmCollectionType {
+    if (isDartCoreSet) return RealmCollectionType.set;
+    if (isDartCoreList) return RealmCollectionType.list;
+    if (isDartCoreMap &&
+        (this as ParameterizedType).typeArguments.first ==
+            session.typeProvider.stringType) {
+      return RealmCollectionType.dictionary;
+    }
+    return RealmCollectionType.none;
+  }
+
+  DartType get basicType {
+    if (isDynamic) return this;
+    if (isNullable) return asNonNullable.basicType;
+    if (isRealmCollection) {
+      return (this as ParameterizedType).typeArguments.last;
+    }
+    if (isRealmModel) {
+      // convert _T to T .. I think I need to implement ClassTypeMacro
```

Comment:
The code works without changes, due to a later hack using replaceAll. I will update the comment. 

The problem is that given `_A` we cannot get `DartType` for `A` since it is not defined yet. The ClassTypeMacro I refer to, is a system coming with the new static meta-programming framework that Google is working on.

We should revisit when that is released.

Reply:
thanks. I guess we can create a TODO here but not sure if it makes sense since when static meta programming is available we will probably have a totally different implementation. In that case probably we don't need a TODO in the comments but an issue in GH.

Reply:
I have added both

Reply:
Fixed